geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/Z2Scheme.scala

```diff
@@ -8,33 +8,64 @@
 
 package org.locationtech.geomesa.fs.storage.common.partitions
 
-import java.util.{Collections, Optional}
 import java.util.regex.Pattern
+import java.util.{Collections, Optional}
 
 import com.vividsolutions.jts.geom.{Geometry, Point}
-import org.locationtech.geomesa.curve.Z2SFC
+import org.locationtech.geomesa.curve.{XZ2SFC, Z2SFC}
 import org.locationtech.geomesa.filter.{FilterHelper, FilterValues}
 import org.locationtech.geomesa.fs.storage.api.{PartitionScheme, PartitionSchemeFactory}
+import org.locationtech.geomesa.fs.storage.common.partitions.SpatialPartitionSchemeConfig._
 import org.locationtech.geomesa.utils.geotools.{GeometryUtils, WholeWorldPolygon}
 import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
 import org.opengis.filter.Filter
 
-class Z2Scheme(bits: Int, geom: String, leaf: Boolean) extends PartitionScheme {
+class Z2Scheme(bits: Int, geom: String, leaf: Boolean) extends SpatialScheme(bits, geom, leaf) {
+  override protected val digits: Int = math.ceil(bits * math.log10(2)).toInt
 
-  require(bits % 2 == 0, "Resolution must be an even number")
+  override def getName: String = Z2Scheme.Name
 
-  // note: z2sfc resolution is per dimension
   private val z2 = new Z2SFC(bits / 2)
-  private val digits = math.ceil(math.log10(math.pow(2, bits))).toInt
-
-  override def getName: String = Z2Scheme.Name
 
   override def getPartition(feature: SimpleFeature): String = {
-    // TODO support non-point geoms
     val pt = feature.getAttribute(geom).asInstanceOf[Point]
     z2.index(pt.getX, pt.getY).z.formatted(s"%0${digits}d")
   }
 
+  override def generateRanges(xy: Seq[(Double, Double, Double, Double)]): Seq[Long] =
+    z2.ranges(xy).flatMap(ir => ir.lower to ir.upper)
+}
+
+class XZ2Scheme(bits: Int, geom: String, leaf: Boolean) extends SpatialScheme(bits, geom, leaf) {
+  // The max XZ2 value is (4^((bits/2)+1)-1)/3.  This calculates the number of digits in that value.
+  override protected val digits: Int = math.ceil(((bits/2)+1)* math.log10(4) - math.log10(3)).toInt
+
+  override def getName: String = XZ2Scheme.Name
+
+  private val xz2 = XZ2SFC((bits / 2).asInstanceOf[Short])
+
+  override def getPartition(feature: SimpleFeature): String = {
+    // TODO support non-point geoms
+    val geometry = feature.getAttribute(geom).asInstanceOf[Geometry]
+    val envelope = geometry.getEnvelopeInternal
+    xz2.index(envelope.getMinX, envelope.getMinY, envelope.getMaxX, envelope.getMaxY).formatted(s"%0${digits}d")
+  }
+
+  override def generateRanges(xy: Seq[(Double, Double, Double, Double)]): Seq[Long] =
+    xz2.ranges(xy).flatMap(ir => ir.lower to ir.upper)
+}
+
+abstract class SpatialScheme(bits: Int, geom: String, leaf: Boolean) extends PartitionScheme {
+
+  require(bits % 2 == 0, "Resolution must be an even number")
+
+  // TODO: The number of digits might be off for Z2 versus XZ2
+  protected val digits: Int // = math.ceil(math.log10(math.pow(2, bits))).toInt
```

Comment:
I assume this comment is no longer relevant and the commented out line can be removed?